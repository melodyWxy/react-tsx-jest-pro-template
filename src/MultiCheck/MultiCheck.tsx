import './MultiCheck.css';

import React, { useState, useMemo, useCallback } from 'react';

export type Option = {
  label: string,
  value: string
}

/**
 * Notice:
 * 1. There should be a special `Select All` option with checkbox to control all passing options
 * 2. If columns > 1, the options should be placed from top to bottom in each column
 *
 * @param {string} label - the label text of this component
 * @param {Option[]} options - options
 * @param {string[]} values -  checked option values
 * @param {string[]} defaultValues - default checked option values
 * @param {number} columns - default value is 1
 * @param {Function} onChange - when checked options are changed,
 *                             they should be passed to outside
 */
type Props = {
  label?: string,
  options: Option[],
  columns?: number,
  values?: string[],
  defaultValues?: string[],
  onChange?: (options: Option[]) => void,
}

const MULTICHECK_SELECTED_ALL = 'MULTICHECK_SELECTED_ALL';
const MULTICHECK_SELECTED_ALL_OPTION = { 
  label: 'Selected All', 
  value: MULTICHECK_SELECTED_ALL
}

const MultiCheck: React.FunctionComponent<Props> = (props: Props): JSX.Element => {
  const { label, options, columns = 1, values, defaultValues, onChange } = props;
  
  // State
  const [ selectedValues, setSelectedValues ] = useState( defaultValues || [] );
  
  // memo-callback
  // getRenderValues-fn
  const getRenderValues = useCallback((currentValues: string[], options: Option[])=>{
    let isSelectedAll = true;
    options.forEach(option => {
      if(!currentValues.find(value=> value === option.value)){
        isSelectedAll = false;
      }
    })
    if(isSelectedAll){
      return [MULTICHECK_SELECTED_ALL].concat(currentValues);
    }
    return currentValues;
  },[])

  // handleCheckedChange
  const handleCheckedChange = (value: string) => {
    let currentValues = values || selectedValues;
    const currentSelectedOptions: Option[] = [];
    if(value === MULTICHECK_SELECTED_ALL) {
      // 全选状态改变单独处理
      if(currentValues.length === options.length){
        currentValues = [];
      }else{
        currentValues = options.map((option: Option) => option.value);
      }
    }else{
      const valueKey = currentValues.indexOf(value);
      if(valueKey === -1){
        currentValues.push(value);
      }else{
        currentValues.splice(valueKey, 1);
      }
    }
    currentValues.forEach((selectedOptionValue: string)=>{
      if(selectedOptionValue !== MULTICHECK_SELECTED_ALL){
        const findOne = options.find((option: Option)=>option.value === selectedOptionValue);
        if(findOne){
          currentSelectedOptions.push(findOne);
        }
      }
    })
    if(!values){
      setSelectedValues([...currentValues]);
    } 
    onChange && onChange(currentSelectedOptions)
  }

  // memo-ui
  // render-labelUI
  const labelUI = useMemo(() => {
    if( label || label === '' ){
      return (
        <div className='MultiCheck-label-wrap'>
          {label}
        </div>
      );
    }
    return null;
  }, [label])

  // render-options
  const optionsUI = useMemo(() => {
    const allOptions = [MULTICHECK_SELECTED_ALL_OPTION].concat(options);
    const currentValues = getRenderValues(values || selectedValues, options);
    const mapOptionColumnList = [];
    const subGroupLength = Math.ceil(allOptions.length/columns);
    let endIndex = 0;
    for(let i = 0; i < allOptions.length; i = endIndex){
       endIndex = i + subGroupLength;
       if(endIndex > allOptions.length){
        endIndex = allOptions.length;
       }
       mapOptionColumnList.push(allOptions.slice(i, endIndex));
    }
    const allItems = mapOptionColumnList.map((columnItem, columnIndex) => {
      const optionsItems = columnItem.map(item => {
        const { label, value } = item;
        return (
          <div className='MultiCheck-option-wrap' key={value}>
            <div>
              <input 
                type='checkbox' 
                value={value} 
                checked={currentValues.includes(value)}
                onChange={()=>handleCheckedChange(value)}
              />
            </div>
            <div>
              { label }
            </div>
          </div>
        )
      })
      return (
        <div className="MultiCheck-column-wrap" key={columnIndex}>
          { optionsItems }
        </div>
      )
    });
    return allItems;
  }, [columns, options, values, selectedValues])

  return (
    <div className='MultiCheck' >
      {labelUI}
      <div className='MultiCheck-content-wrap'>
        <div className='MultiCheck-content'>
          {optionsUI}
        </div>
      </div>
    </div>
  )
}

export default MultiCheck;
