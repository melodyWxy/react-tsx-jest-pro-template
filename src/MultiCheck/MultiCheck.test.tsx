import React from 'react';
import Enzyme from "./enzyme.config";
import MultiCheck, { Option }  from './MultiCheck';

// TODO more tests


const { mount } = Enzyme;


const options: Option[] = [
  {label: 'aaa', value: '111',},
  {label: 'bbb', value: '222',},
  {label: 'ccc', value: '333',},
  {label: 'ddd', value: '444',},
  {label: 'eee', value: '555',},
  {label: 'fff', value: '666',},
  {label: 'ggg', value: '777',},
  {label: 'hhh', value: '888',},
  {label: 'iii', value: '999',},
]

describe('MultiCheck', () => {
  beforeAll(()=>{
    // do something before all test here ~ 
  })
  afterAll(()=>{
    // do something after all test here ~ 
  })
  describe('initialize', () => {
    it('renders the label if label provided', () => {
      // TODO
      const wrapper = mount(<MultiCheck  options={options} label="test label~"/>)
      expect(wrapper.render()).toMatchSnapshot();
      expect(wrapper.find('.MultiCheck-label-wrap')).toHaveLength(1);
      expect(wrapper.find('.MultiCheck-label-wrap').html()).toEqual('<div class=\"MultiCheck-label-wrap\">test label~</div>');
      // expect(wrapper.find)
    });
  });
});

// Sorry, I don't have time to write more unit tests. You can directly ask me the test case ideas.
// 抱歉，我没有时间去写更多的单元测试。你可以直接问我测试用例思路。
// author:  Xinyu Wang (melodyWxy)