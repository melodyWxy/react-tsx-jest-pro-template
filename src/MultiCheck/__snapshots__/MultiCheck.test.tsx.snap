// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiCheck initialize renders the label if label provided 1`] = `
<div
  class="MultiCheck"
>
  <div
    class="MultiCheck-label-wrap"
  >
    test label~
  </div>
  <div
    class="MultiCheck-content-wrap"
  >
    <div
      class="MultiCheck-content"
    >
      <div
        class="MultiCheck-column-wrap"
      >
        <div
          class="MultiCheck-option-wrap"
        >
          <div>
            <input
              type="checkbox"
              value="MULTICHECK_SELECTED_ALL"
            />
          </div>
          <div>
            Selected All
          </div>
        </div>
        <div
          class="MultiCheck-option-wrap"
        >
          <div>
            <input
              type="checkbox"
              value="111"
            />
          </div>
          <div>
            aaa
          </div>
        </div>
        <div
          class="MultiCheck-option-wrap"
        >
          <div>
            <input
              type="checkbox"
              value="222"
            />
          </div>
          <div>
            bbb
          </div>
        </div>
        <div
          class="MultiCheck-option-wrap"
        >
          <div>
            <input
              type="checkbox"
              value="333"
            />
          </div>
          <div>
            ccc
          </div>
        </div>
        <div
          class="MultiCheck-option-wrap"
        >
          <div>
            <input
              type="checkbox"
              value="444"
            />
          </div>
          <div>
            ddd
          </div>
        </div>
        <div
          class="MultiCheck-option-wrap"
        >
          <div>
            <input
              type="checkbox"
              value="555"
            />
          </div>
          <div>
            eee
          </div>
        </div>
        <div
          class="MultiCheck-option-wrap"
        >
          <div>
            <input
              type="checkbox"
              value="666"
            />
          </div>
          <div>
            fff
          </div>
        </div>
        <div
          class="MultiCheck-option-wrap"
        >
          <div>
            <input
              type="checkbox"
              value="777"
            />
          </div>
          <div>
            ggg
          </div>
        </div>
        <div
          class="MultiCheck-option-wrap"
        >
          <div>
            <input
              type="checkbox"
              value="888"
            />
          </div>
          <div>
            hhh
          </div>
        </div>
        <div
          class="MultiCheck-option-wrap"
        >
          <div>
            <input
              type="checkbox"
              value="999"
            />
          </div>
          <div>
            iii
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
